File: backend/app/schemas/two_factor.py
# app/schemas/two_factor.py

from pydantic import BaseModel
from typing import List


class TwoFactorSetupResponse(BaseModel):
    """Returned when user initiates 2FA setup — frontend renders QR from uri"""
    otpauth_uri: str
    secret: str           # shown as manual fallback (masked in UI)


class TwoFactorConfirmRequest(BaseModel):
    code: str             # 6-digit code from authenticator app


class TwoFactorConfirmResponse(BaseModel):
    message: str
    backup_codes: List[str]  # shown ONCE — user must save these


class TwoFactorVerifyRequest(BaseModel):
    """Sent during login when 2FA is required"""
    session_token: str    # ties this request to the pending login
    code: str             # 6-digit TOTP or 10-digit backup code


class TwoFactorLoginResponse(BaseModel):
    """Returned by login() when 2FA is enabled — not yet fully authenticated"""
    two_factor_required: bool = True
    session_token: str    # opaque token, stored in Redis, ties to user_id

File: frontend/types/two-factor.ts
import { LoginResponse } from "./auth"

export interface TwoFactorSetupResponse {
  otpauth_uri: string
  secret: string
}

export interface TwoFactorConfirmRequest {
  code: string
}

export interface TwoFactorConfirmResponse {
  message: string
  backup_codes: string[]
}

export interface TwoFactorVerifyRequest {
  session_token: string
  code: string
}

export interface TwoFactorLoginResponse {
  two_factor_required: true
  session_token: string
}

/**
 * Type guard — narrows the loginUser thunk's union return type.
 * Needed because LoginResponse and TwoFactorLoginResponse share
 * no common discriminant except this field's presence.
 */
export function isTwoFactorResponse(
  payload: LoginResponse | TwoFactorLoginResponse
): payload is TwoFactorLoginResponse {
  return (payload as TwoFactorLoginResponse).two_factor_required === true
}